QuerySet.annotate improvements

syastrov opened the pull request:
- QuerySet.annotate now returns correct self-type rather than QuerySet[Any]. Attribute access falls back to Any.
- QuerySets that have an annotated model do not report errors during .filter() when called with invalid fields.
- QuerySets that have an annotated model return ordinary dict rather than TypedDict for .values()
- QuerySets that have an annotated model return Any rather than typed Tuple for .values_list()

Calling QuerySet.annotate introduces an "annotated" model type which is a subtype of the original model dynamically (named "MyModel (annotated)"), but which allows arbitrary getting/setting of attributes.

This is originally to fix this issue:

```
a = MyModel.objects.filter()
a = a.annotate() 
```

error: Incompatible types in assignment (expression has type “QuerySet[Any]“, variable has type “Manager[Any]“)

The problem is because of not using self-type

syastrov committed:
QuerySet.annotate returns self-type. Attribute access falls back to Any.

- QuerySets that have an annotated model do not report errors during .filter() when called with invalid fields.
- QuerySets that have an annotated model return ordinary dict rather than TypedDict for .values()
- QuerySets that have an annotated model return Any rather than typed Tuple for .values_list()

sobolevn reviewed (dismissed):
Great work!

The only question I have is: how one can represent annotated `User` statically? For example, we have this code:

```python
def allowed(param: User):
     return param.foo  # E: "User" has no attribute "foo" 
```

But, when one will pass there an instance with `.annotate(foo=...)` it will work in runtime, but fail in typechecking. Do you have any solution in mind?

syastrov committed:
Fix .annotate so it reuses existing annotated types. Fixes error in typechecking Django testsuite.

syastrov committed:
Fix self-typecheck error

syastrov committed:
Fix flake8

syastrov committed:
Fix case of .values/.values_list before .annotate.

syastrov committed:
Extra ignores for Django 2.2 tests (false positives due to tests assuming QuerySet.first() won't return None)

Fix mypy self-check.

syastrov committed:
More tests + more precise typing in case annotate called before values_list.

Cleanup tests.

syastrov committed:
Test and fix annotate in combination with values/values_list with no params.

syastrov committed:
Remove line that does nothing :)

kszmigiel reviewed (approved):
Looks totally good for me
@syastrov thank you for you work!
@sobolevn How should we deal with all the changes in mypy plugin created since the refactor branch was created? Should I apply all these changes to the refactor branch by hand? What's your opinion?

syastrov committed:
Formatting fixes

sobolevn commented on mypy_django_plugin/transformers/orm_lookups.py:
```diff
@@ -29,7 +30,11 @@ def typecheck_queryset_filter(ctx: MethodContext, django_context: DjangoContext)
-        lookup_type = django_context.resolve_lookup_expected_type(ctx, model_cls, lookup_kwarg)
+        lookup_type: MypyType
+        if is_annotated_model_fullname(model_cls_fullname):
+            lookup_type = AnyType(TypeOfAny.explicit)
```
I prefer `.implementation_detail`, because `.explicit` can raise errors if `no-explicit-any` is set.

sobolevn commented on mypy_django_plugin/transformers/querysets.py:
```diff
@@ -38,12 +42,19 @@ def determine_proper_manager_type(ctx: FunctionContext) -> MypyType:
+    *,
+    method: str,
+    lookup: str,
+    silent_on_error: bool = False
```
`False,`

sobolevn commented on mypy_django_plugin/transformers/querysets.py:
```diff
@@ -61,7 +72,12 @@ def get_field_type_from_lookup(
-    ctx: MethodContext, django_context: DjangoContext, model_cls: Type[Model], flat: bool, named: bool
+    ctx: MethodContext,
+    django_context: DjangoContext,
+    model_cls: Type[Model],
```
Let's make all bool args `kw-only` args. Anyway, we use them as such! 👍

sobolevn commented on mypy_django_plugin/transformers/querysets.py:
```diff
@@ -145,10 +175,69 @@ def extract_proper_type_queryset_values_list(ctx: MethodContext, django_context:
+        annotated_type = model_type
+    else:
+        annotated_typeinfo = helpers.lookup_fully_qualified_typeinfo(
+            cast(TypeChecker, api), model_module_name + "." + type_name
```
If we already use `cast` here and declare `api` as `api: CheckerPluginInterface`, maybe we can change the definition?

syastrov replied:
The definition comes from `ctx` which is `MethodContext` (defined by mypy) as `CheckerPluginInterface`.
I only annotated it locally (redundantly) as `CheckerPluginInterface` since my IDE couldn't figure it out for some reason.
I'll remove the redundant annotation.

sobolevn commented on tests/typecheck/managers/querysets/test_annotate.yml:
```diff
@@ -0,0 +1,207 @@
+        from django.db.models.expressions import F
+
+        qs = User.objects.annotate(foo=F('id'))
+        reveal_type(qs) # N: Revealed type is 'django.db.models.manager.Manager[myapp.models.User (annotated)]'
```
Quick question: can we use the annotation type in the type's repr? Would it be more or less readable / useful?

syastrov replied:
Not sure what you mean by "the annotation type". Is that `foo`? It would be cool to know what it was annotated with, but that's outside the scope of this PR.
Anyway, from my experimentation, we cannot put anything with `.` in the type's name or it will interfere with lookups.

sobolevn commented on tests/typecheck/managers/querysets/test_annotate.yml:
```diff
@@ -0,0 +1,207 @@
+        qs = User.objects.annotate(foo=F('id'))
+        qs = qs.annotate(bar=F('id'))
+        annotated = qs.get()
+        reveal_type(annotated.foo) # N: Revealed type is 'Any'
```
What about `annotated.bar` here?

syastrov replied:
Added

sobolevn reviewed (commented):
This is a great piece of work! 🏆 
Thanks a lot!

syastrov committed:
Merge branch 'master' into queryset-annotate

syastrov committed:
Address code review

syastrov committed:
Fix quoting in tests after mypy changed things

sobolevn commented on mypy_django_plugin/lib/constants.py:
```diff
@@ -0,0 +1 @@
+ANNOTATED_SUFFIX = " (annotated)"
```
`: Final`?

sobolevn reviewed (approved):
Thanks a lot, @syastrov!

You provide such great pieces of code for this project! I really appreciate it 👍

syastrov committed:
Use Final

sobolevn commented on mypy_django_plugin/lib/constants.py:
```diff
@@ -0,0 +1,3 @@
+from typing import Final
```
Oups, it is `typing_extensions` for `python<3.8`

syastrov committed:
Use typing_extensions.Final

syastrov committed:
Merge branch 'master' into queryset-annotate

syastrov committed:
Fixes after ValuesQuerySet -> _ValuesQuerySet refactor. Still not passing tests yet.

syastrov committed:
Fix inheritance of _ValuesQuerySet and remove unneeded type ignores.

This allows the test
"annotate_values_or_values_list_before_or_after_annotate_broadens_type"
to pass.

syastrov committed:
Merge branch 'master' into queryset-annotate

At the same time, remove added _ValuesQuerySet methods, since they are
not needed after removing type ignore on the line declaring the class.

This also corrects the _Row type of .annotate, turning it from Any to
tuple[Any].

syastrov committed:
Make it possible to annotate user code with "annotated models", using PEP 583 Annotated type.

syastrov committed:
Add docs

sobolevn commented on README.md:
```diff
@@ -179,6 +179,28 @@ def use_my_model():
+    username = models.CharField(max_length=100)
+
+
+def func(m: WithAnnotations[MyModel]) -> str:
```
Will it be possible for us to later add something like `WithAnnotations[Model, TypedDict]` and not break things for users? Double checking 🙂

sobolevn commented on mypy_django_plugin/django/context.py:
```diff
@@ -113,7 +115,12 @@ def model_modules(self) -> Dict[str, Set[Type[Model]]]:
     def get_model_class_by_fullname(self, fullname: str) -> Optional[Type[Model]]:
-        # Returns None if Model is abstract
+        """Returns None if Model is abstract"""
+        ANNOTATED_PREFIX: Final = WITH_ANNOTATIONS_FULLNAME + "["
```
No need to make it `Final` and constant inside the function.

sobolevn commented on mypy_django_plugin/lib/helpers.py:
```diff
@@ -61,7 +62,15 @@ def is_toml(filename: str) -> bool:
     if "." not in fullname:
         return None
-    module, cls_name = fullname.rsplit(".", 1)
+    if "[" in fullname:
```
`if "[" in fullname and "]" in fullname:`?

sobolevn commented on mypy_django_plugin/transformers/models.py:
```diff
@@ -390,3 +393,59 @@ def set_auth_user_model_boolean_fields(ctx: AttributeContext, django_context: Dj
+
+
+def get_or_create_annotated_type(
+    api: Union[SemanticAnalyzer, CheckerPluginInterface], model_type: Instance
```
`api: SemanticAnalyzer`?

sobolevn replied:
Or even `TypeAnalyzer`

syastrov replied:
It's unfortunately used both in `MethodContext` (`CheckerPluginInterface`) and `AnalyzeTypeContext` (where I can grab `SemanticAnalyzer` from the `api.api` attribute).

sobolevn reviewed (commented):
Great work. Like really great work! 👍

syastrov committed:
Make QuerySet[_T] an external alias to _QuerySet[_T, _T].

This currently has the drawback that error messages display the internal type _QuerySet, with both type arguments.

See also discussion on #661 and #608.

Fixes #635: QuerySet methods on Managers (like .all()) now return QuerySets rather than Managers.

Address code review by @sobolevn.

sobolevn reviewed (approved):
This is pure gold 🏅 Thank you!

Merge when you fill confident!

syastrov committed:
Support passing TypedDicts to WithAnnotations

sobolevn commented on README.md:
```diff
@@ -179,6 +179,47 @@ def use_my_model():
+    foo: str
+
+def func2(m: WithAnnotations[MyModel, MyTypedDict]) -> str:
+    return m.foo # OK, since we said field "foo" was allowed
```
Let's show `m.bar` example as well, it should fail, doesn't it?

syastrov replied:
Sure, added it. Yes it would fail and there is a test for it already in `annotate_using_with_annotations_typeddict`.

syastrov committed:
Add an example of an error to README regarding WithAnnotations + TypedDict.

sobolevn reviewed (approved):
💯

syastrov committed:
Fix runtime behavior of ValuesQuerySet alias (you can't extend Any, for example).

Fix some edge case with from_queryset after QuerySet changed to be an
alias to _QuerySet. Can't make a minimal test case as this only occurred
on a large internal codebase.

syastrov committed:
Fix issue when using from_queryset in some cases when having an argument with a type annotation on the QuerySet.

The mypy docstring on anal_type says not to call defer() after it.
